python/src/iceberg/utils/schema_conversion.py

```diff
@@ -0,0 +1,451 @@
+            ...     ),
+            ...     NestedField(
+            ...         field_id=501, name="manifest_length", field_type=LongType(), is_optional=False, doc="Total file size in bytes"
+            ...     ),
+            ...     schema_id=1
+            ... )
+            >>> avro_schema == iceberg_schema
+            True
+
+        Args:
+            avro_schema (Dict[str, Any]): The JSON decoded Avro schema
+
+        Returns:
+            Equivalent Iceberg schema
+        """
+        return Schema(*[self._convert_field(field) for field in avro_schema["fields"]], schema_id=1)
+
+    def _resolve_union(self, type_union: Dict | List | str) -> Tuple[str | Dict[str, Any], bool]:
+        """
+        Converts Unions into their type and resolves if the field is optional
+
+        Examples:
+            >>> AvroSchemaConversion()._resolve_union('str')
+            ('str', False)
+            >>> AvroSchemaConversion()._resolve_union(['null', 'str'])
+            ('str', True)
+            >>> AvroSchemaConversion()._resolve_union([{'type': 'str'}])
+            ({'type': 'str'}, False)
+            >>> AvroSchemaConversion()._resolve_union(['null', {'type': 'str'}])
+            ({'type': 'str'}, True)
+
+        Args:
+            type_union: The field, can be a string 'str', list ['null', 'str'], or dict {"type": 'str'}
+
+        Returns:
+            A tuple containing the type and nullability
+
+        Raises:
+            TypeError: In the case non-optional union types are encountered
+        """
+        avro_types: Dict | List
+        if isinstance(type_union, str):
+            # It is a primitive and required
+            return type_union, False
+        elif isinstance(type_union, dict):
+            # It is a context and required
+            return type_union, False
+        else:
+            avro_types = type_union
+
+        is_optional = "null" in avro_types
+
+        if len(avro_types) > 2:
+            raise TypeError("Non-optional types aren't part of the Iceberg specification")
+
+        # Filter the null value and return the type
+        return list(filter(lambda t: t != "null", avro_types))[0], is_optional
+
+    def _resolve_inner_type(self, avro_type: Dict[str, Any], inner_field_name: str, id_field: str) -> Tuple[IcebergType, bool]:
+        """
+        Removes any unions inside an array or map, and appends the field-id downstream (if required)
+
+        Args:
+            avro_type: The parent Avro type
+            inner_field_name: For example 'items' in case of an array
+            id_field: Contains the element or key/value id, for example element-id, to be copied to the field-id
+
+        Returns:
+            The inner type and the nullability
+        """
+        plain_type, element_is_optional = self._resolve_union(avro_type[inner_field_name])
+        if isinstance(plain_type, dict):
+            # We need the element-id downstream
+            plain_type["field-id"] = avro_type[id_field]
```

Comment:
I strongly prefer not modifying the incoming data like this is doing. I think `avro_type` and parts of that type should be handled as though they are immutable.

Also, I don't think that this works. `_convert_schema` is going to return a type. For primitive types, there is no way to pass in the `field-id` property, and in all cases the `field-id` is going to be ignored because Iceberg types don't carry field IDs.

I think the simple solution to both problems is to handle inner types in the map and array methods directly, rather than using a common method.

Reply:
Hey Ryan, I agree. I don't think we really need the method at all. I've removed it and just resolved the type instead of the field. Still some old code dangling around.